09-features.md

```diff
@@ -41,6 +41,7 @@ The Context column decodes as follows:
 | 20/21 | `option_anchor_outputs`          | Anchor outputs                                            | IN       | `option_static_remotekey` | [BOLT #3](03-transactions.md)         |
 | 22/23 | `option_anchors_zero_fee_htlc_tx` | Anchor commitment type with zero fee HTLC transactions   | IN       |                   | [BOLT #3][bolt03-htlc-tx], [lightning-dev][ml-sighash-single-harmful]|
 | 26/27 | `option_shutdown_anysegwit`         | Future segwit versions allowed in `shutdown`              | IN       |                   | [BOLT #2][bolt02-shutdown]   |
+| 48    | `option_payment_metadata` | Payment metadata in tlv record | 9 | | [BOLT #11](11-payment-encoding.md#tagged-fields)
```

Comment:
Grabbed feature bit 48 for this.

Reply:
This should have an optional bit too, no? And the category is missing.

Reply:
My thought was that if an invoice specifies metadata, you have to include it in the onion as a sender. Therefore a required invoice feature. We can add an optional bit, but is there a use case for that? A use case where including the metadata is somehow better for the sender, but it will also work without?

Reply:
Yea, I don't see why we should forbid using metadata as an optional thing. We could debate whether we should set a feature bit for that vs just have the field in the invoice, but why would we want to forbid such a use?

Reply:
My general answer would be because every variation has a development and maintenance cost.

Maybe it is worth thinking a little harder about the actual use case. An invoice with optional metadata. New implementations will send the metadata, but older ones will not. The receiver needs that metadata, but is also able to handle a payment without. How can this be used?

Reply:
I have a different argument for allowing the optional bit number.

If we don't allow the optional bit to be set, we cannot monitor whether payers have upgraded or not, so we will never have enough data to know whether we can turn it on with the mandatory bit. We will be forced to just decide at some point to make it mandatory in a specific release, and realize afterwards that it is actually breaking a lot of payments.

The way I plan on using it in eclair is that in the initial release, the optional bit will be set and a small, dummy metadata field will be included in every invoice. This way receivers can track what percentage of their payers actually support this option (by checking whether they included the metadata in the payment or not), and when it reaches 100% (or a good enough threshold) they can decide to make it mandatory, knowing that it will not affect their existing customers.

Does that sound reasonable?

Reply:
> My general answer would be because every variation has a development and maintenance cost.

Hmm, but I'm not even sure what the variation here is? As far as I'm aware, all implementations today have some logic to parse feature bits and reject things if there's unknown even bits, and ignore odd bits. This bit being odd should be +0-0 in terms of LoC, at least it would be in LDK.

I guess I don't understand what your view is on optional use (and maybe the spec needs clarification) - is it okay to include a metadata field and no feature bit (and should the sender, then, include the metadata if they understand it?)?

> I have a different argument for allowing the optional bit number.

It seems like this is doable with or without the feature bit? You can still track the number of senders that included the new data.

Reply:
>The way I plan on using it in eclair is that in the initial release, the optional bit will be set and a small, dummy metadata field will be included in every invoice. This way receivers can track what percentage of their payers actually support this option

Interesting use case for optional. I hadn't thought about this before, but indeed, it is a critical that a large majority has upgraded before switching over to stateless invoices.

Reply:
> It seems like this is doable with or without the feature bit? You can still track the number of senders that included the new data.

True, but it requires you to include a `payment_metadata` in the invoice without activating the feature bit, which is a bit less clean in my opinion, if we have a feature bit for this, we should use it.

Reply:
Agreed. We would need to implement optimistic metadata sending without feature bit, not the cleanest logic.

Reply:
Added optional bit